File: Source/CSharp/Expressive/Expressive/Expressions/VariableExpression.cs
using System;
using System.Collections.Generic;

namespace Expressive.Expressions
{
    internal class VariableExpression : IExpression
    {
        private readonly string variableName;

        internal VariableExpression(string variableName)
        {
            this.variableName = variableName;
        }

        #region IExpression Members

        public object Evaluate(IDictionary<string, object> variables)
        {
            if (variables == null ||
                !variables.ContainsKey(this.variableName))
            {
                throw new ArgumentException("The variable '" + this.variableName + "' has not been supplied.");
            }

            // Check to see if we have to referred to another expression.
            var expression = variables[this.variableName] as Expression;
            if (expression != null)
            {
                return expression.Evaluate(variables);
            }

            return variables[this.variableName];
        }

        #endregion
    }
}

File: Source/Java/Expressive/src/main/java/com/bijington/expressive/expressions/VariableExpression.java
package com.bijington.expressive.expressions;

import com.bijington.expressive.Expression;
import com.bijington.expressive.exceptions.ExpressiveException;

import java.util.Map;

/**
 * Created by shaun on 05/07/2016.
 */
public class VariableExpression implements IExpression {
    private final String _variableName;

    public VariableExpression(String variableName) {
        _variableName = variableName;
    }

    @Override
    public Object evaluate(Map<String, Object> variables) throws ExpressiveException {
        if (variables == null ||
            !variables.containsKey(_variableName))
        {
            throw new ExpressiveException("The variable '" + _variableName + "' has not been supplied.");
        }

        Object value = variables.get(_variableName);
        // Check to see if we have to referred to another expression.
        if (value != null && value.getClass().equals(Expression.class))
        {
            Expression expression = (Expression)value;

            return expression.evaluate(variables);
        }

        return value;
    }
}
